Modules/itertoolsmodule.c

```diff
@@ -142,35 +142,35 @@ static PyObject *
 batched_next(batchedobject *bo)
 {
     Py_ssize_t i;
+    Py_ssize_t n = bo->batch_size;
     PyObject *it = bo->it;
     PyObject *item;
     PyObject *result;
 
     if (it == NULL) {
         return NULL;
     }
-    result = PyList_New(0);
+    result = PyList_New(n);
     if (result == NULL) {
         return NULL;
     }
-    for (i=0 ; i < bo->batch_size ; i++) {
+    for (i=0 ; i < n ; i++) {
         item = PyIter_Next(it);
         if (item == NULL) {
             break;
```

Comment:
Do we care how this handles strange iterators that keep going after having raised `StopIteration`?

```
>>> from itertools import batched
>>> class A:
...     def __init__(self):
...         self.i = 0
...     def __iter__(self):
...         return self
...     def __next__(self):
...         i = self.i
...         self.i += 1
...         if i == 7:
...             raise StopIteration()
...         return i
...
>>> x = batched(A(), 5)
>>> next(x)
[0, 1, 2, 3, 4]
>>> next(x)
[5, 6]
>>> next(x)
[8, 9, 10, 11, 12]
```

Maybe the `Py_CLEAR(bo->it)` could be moved to before this `break;`.

Reply:
Yes.  The batched iterator should permanently terminate for both the empty list path and the short list path.

Reply:
This does seem like an edge case that could go either way.

I'm noticing that some iterators do `Py_CLEAR(obj->it)` as soon as the `next()` call fails, while other objects have internal iterators that are never NULL. My concern is that this clears the iterator sometimes (when the first `next()` of a batch fails), but not other times (when a subsequent `next()` fails).

So I would think to do one or the other, but not both:

1. Clear the iterator immediately after *any* internal `next()` call fails, similar to `pairwise`, `islice`, `chain`, and `cycle` (sort of).

2. Never clear the iterator and remove the `if (it == NULL)` check altogether, similar to `dropwhile`, `takewhile`, `accumulate`, `compress`, `zip` and `enumerate`. Unfortunately, it would mean allocating and throwing away `PyList_New(n)`, but that's in the rare case that someone is abusing the iterator.

Maybe removing the `Py_CLEAR` and the `NULL` check is most natural?